More coccinelle fixes

qmonnet opened the pull request:
- bpf: Add semi-colons after __maybe_unused label attributes
- bpf: Add some missing const qualifiers
- cocci: Set COCCINELLE_HOME to fix path for isomorphisms and macros files

Please look at commit descriptions for details.

```release-note
cocci: Work around a bug in coccinelle to better check files, add a few missing `const` qualifiers to BPF code
```

qmonnet committed:
bpf: Add some missing const qualifiers

All missing qualifiers detected by coccinelle.

The missing "const" on the "mac" argument in icmp6.h was found by
running the latest coccinelle (not 1.1.1 as in CI and Docker image) on
the bpf code base.

The missing qualifier in proxy.h was not detected so far because
coccinelle stumbles on parsing macro CTX_REDIRECT_FN in proxy.h, in
particular because of the CT_TUPLE_TYPE argument where we pass a
two-word long type (struct ipv(4|)_ct_tuple). The spatch tool would fail
with:

    EXN: Failure("bpf/lib/proxy.h: 164: try to delete an expanded token: struct") in bpf/lib/proxy.h

and then would fail to detect the missing "const". By commenting out
temporarily the calls to CTX_REDIRECT_FN we make it work again. Good
news, by adding the missing "const" to "CT_TUPLE_TYPE", we're also
fixing the parsing for coccinelle, so future mistakes in the file should
be reported as expected.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
cocci: Set COCCINELLE_HOME to fix path for isomorphisms and macros files

Trying to create a new coccinelle check, we realised that some patterns
that we expected to match in the bpf/ were in fact never caught.

Here is a minimal reproducer.

foo.cocci:

    @@
    @@
    (
    - foo = bar;
    + foo = baz;
    )

foo.c (not working):

    int main (void)
    {
    	int foo, bar, baz = 0, coccibreaker __maybe_unused;

    	foo = bar;
    	return foo;
    }

Invocation:

    $ docker run --rm --user 1000 \
        --workdir /workspace -v `pwd`:/workspace -it
        docker.io/cilium/coccicheck:2.4@sha256:24abe3fbb8e829fa41a68a3b76cb4df84fd5a87a7d1d6254c1c1fe5effb5bd1b \
        spatch --include-headers --very-quiet --in-place \
        --sp-file contrib/coccinelle/foo.cocci bpf/foo.c

This does not catch the "foo = bar;" statement in the C file.
However, the pattern is caught in the following file instead:

foo.c (working):

    int main (void)
    {
    	int foo, bar, baz = 0;
    	int coccifine __maybe_unused;

    	foo = bar;
    	return foo;
    }

The __maybe_unused attribute seems to prevent spatch to parse correctly
the function. This happens independently of whether __maybe_unused is
defined in the C code. The behaviour remains the same if we replace
__maybe_unused with some other random word (although replacing it with
"__attribute__((unused))" works as expected).

Running from coccinelle's "master" branch works fine, and when running
spatch locally (no Docker image) we can also observe the following
warnings:

    warning: Can't find macro file: /usr/local/bin/lib/coccinelle/standard.h
    warning: Can't find default iso file: /usr/local/bin/lib/coccinelle/standard.iso

This error was fixed with commit coccinelle/coccinelle@540888ff426e
("Fix 263: wrong default path for COCCINELLE_HOME") [0], which is not
part of a release yet.

Until it is, we still want to have our code covered by coccinelle. As a
workaround to the issue, we could pass manually the paths to the files
containing the definitions for the isomorphisms and built-in macros for
coccinelle, with spatch options "--macro-file-builtin" and "--iso-file".
Even simpler, we can set $(COCCINELLE_HOME) to the correct path so that
both files are found.

No camels were harmed during this investigation.

[0] https://gitlab.inria.fr/coccinelle/coccinelle/-/commit/540888ff426e

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

kaworu commented on bpf/Makefile.bpf:
```diff
@@ -77,9 +77,15 @@ checkpatch:
 coccicheck:
 	$(QUIET) $(foreach TARGET,$(shell find $(ROOT_DIR)/contrib/coccinelle/ -name '*.cocci'), \
-		spatch --no-show-diff --sp-file ${TARGET} --include-headers --very-quiet --dir . ; )
+		spatch --no-show-diff --sp-file ${TARGET} --include-headers --very-quiet \
```
nit: avoid mixing `$()` and `${}`
```suggestion
		spatch --no-show-diff --sp-file $(TARGET) --include-headers --very-quiet \
```

pchaigno replied:
Only nits and a reject review. You're a harsh reviewer! :smile_cat:

qmonnet replied:
:)

I don't know, “Changes requested” means we expect the PR to be updated, I don't find it harsh. Makes reviewer's intention clear, at least.

Then of course I'd adjust depending on who is the author, it's true the “rejection” aspect might discourage newer contributors.

kaworu replied:
@pchaigno for context at the time I reviewed CI didn't pass, if it did I wouldn't have requested changes.

kaworu reviewed (changes requested):
`s/Good new,/Good news,/` in the first commit message, `s/"__attribute__((unused))/"__attribute__((unused))"/` in the second commit message, other than that and the comment about the `make` variable that LGTM.

gentoo-root commented on bpf/Makefile.bpf:
```diff
@@ -77,9 +77,15 @@ checkpatch:
+# Note: --macro-file-builtins and --iso-file can be removed, here and in the
+# messages in the .cocci files, next time we upgrade coccinelle. The issue was
+# fixed in https://gitlab.inria.fr/coccinelle/coccinelle/-/commit/540888ff426e.
+COCCI_PATHS ?= --macro-file-builtins /usr/local/lib/coccinelle/standard.h --iso-file /usr/local/lib/coccinelle/standard.iso
```
Maybe it's easier and less intrusive to use this?

```
export COCCINELLE_HOME ?= /usr/local/lib/coccinelle
```

It's easier for users to override if they installed coccinelle differently (e.g., with a package manager into /usr without /local): just replace a path, no need to type the options. It also doesn't require changing the sparse command below.

gentoo-root replied:
BTW, regarding the upstream [commit](https://gitlab.inria.fr/coccinelle/coccinelle/-/commit/540888ff426e) in coccinelle, does this make sense to anyone?

```
        Str.global_replace (Str.regexp "[$]{exec_prefix}")
          (Filename.dirname bin_dir) "@libdir@"
```

It takes libdir, looks for exec_prefix and replaces it with the parent directory of bindir. What's the point of this transformation, why not take libdir directly? For example, if exec_prefix is /usr, bindir is /usr/bin, libdir is /usr/lib, you could just take libdir, instead of replacing /usr (parent dir of /usr/bin) with /usr (exec_prefix) in "/usr/lib". Moreover, if someone configures less common dirs, like bindir=/bin, exec_prefix=/usr, libdir=/usr/lib, the whole thing will break, because it will look in /usr/usr/lib.

qmonnet replied:
I haven't checked by setting `COCCINELLE_HOME` (couldn't find related documentation - although I didn't spend hours searching). I'll try that, if it works, yes it would be cleaner. Thanks!

gentoo-root replied:
> couldn't find related documentation - although I didn't spend hours searching

Me neither, I just opened the [link that you put into the commit message](https://gitlab.inria.fr/coccinelle/coccinelle/-/commit/540888ff426e) and read the code and comments :D

```
let path =
  (* COCCINELLE_HOME is the directory where standard.iso is looked for. *)
  try (Sys.getenv "COCCINELLE_HOME")
```

I hope that it will work.

qmonnet replied:
I'm just seeing your second comment - yeah, agree that the fix looks somewhat dubious. I found that it fixed the issue we had and didn't really look further to be honest.

qmonnet committed:
bpf: Add semi-colons after __maybe_unused label attributes

GCC and clang support label attributes, such as __attribute__((unused))
to indicate that under certain configurations, (goto) labels will not be
used at compilation time. We've been using this to silence warnings
about unused labels.

However, GCC's "Attribute Syntax" documentation [0] states that:

    GNU C++ only permits attributes on labels if the attribute specifier
    is immediately followed by a semicolon (i.e., the label applies to
    an empty statement). If the semicolon is missing, C++ label
    attributes are ambiguous, as it is permissible for a declaration,
    which could begin with an attribute list, to be labelled in C++.

We're using clang to compile the BPF programs and I couldn't find an
equivalent statement, but it should likely be the same. Let's add the
semi-colons.

The real motivation for this clean-up is to coccinelle parse the files
properly and to better apply checks. We've seen it struggle when the
semi-colons are missing.

[0] https://gcc.gnu.org/onlinedocs/gcc/Attribute-Syntax.html#Label-Attributes-2

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

pchaigno reviewed (approved):
LGTM. Thanks for the very clear commit messages!

kaworu reviewed (approved):
Thanks @qmonnet! TIL about semi-colons after `__maybe_unused` for goto labels, but that make sense.
